packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/bwc_update.ts

```diff
@@ -0,0 +1,320 @@
+}
+
+export const performBWCUpdate = async <T>(
+  { id, type, attributes, options }: PerformUpdateParams<T>,
+  {
+    registry,
+    helpers,
+    allowedTypes,
+    client,
+    serializer,
+    extensions = {},
+    logger,
+  }: ApiExecutionContext
+): Promise<SavedObjectsUpdateResponse<T>> => {
+  const time = getCurrentTime();
+
+  const {
+    common: commonHelper,
+    validation: validationHelper,
+    encryption: encryptionHelper,
+    preflight: preflightHelper,
+    migration: migrationHelper,
+  } = helpers;
+
+  const { securityExtension } = extensions;
+
+  const namespace = commonHelper.getCurrentNamespace(options.namespace);
+  const {
+    version,
+    references,
+    upsert, // translates to overwrite false if defined, otherwise the doc exists and we overwrite it
+    refresh = DEFAULT_REFRESH_SETTING,
+    // retryOnConflict = version ? 0 : DEFAULT_RETRY_COUNT, // not an option for `create/index`
+    migrationVersionCompatibility,
+  } = options;
+
+  let originalDocMigrated: SavedObject<T>;
+  /* ================= GLOBAL VALIDATION ================== */
+  if (!allowedTypes.includes(type)) {
+    throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+  }
+  if (!id) {
+    throw SavedObjectsErrorHelpers.createBadRequestError('id cannot be empty');
+  }
+  /* ================= PREFLIGHT ================= */
+  /* ================= DOC EXISTS VALIDATION: preflight helper ================== */
+
+  logger.info('Checking doc exists');
+
+  const preflightGetDocResult = await preflightHelper.preflightGetDoc({
+    type,
+    id,
+    namespace,
+  });
+
+  if (preflightGetDocResult.rawDocSource && preflightGetDocResult.checkDocFound === 'found') {
+    const originalDoc = getSavedObjectFromSource<T>(
+      registry,
+      type,
+      id,
+      preflightGetDocResult.rawDocSource,
+      {
+        migrationVersionCompatibility,
+      }
+    );
+
+    try {
+      originalDocMigrated = migrationHelper.migrateStorageDocument(originalDoc) as SavedObject<T>;
+    } catch (error) {
+      throw SavedObjectsErrorHelpers.decorateGeneralError(
+        error,
+        'Failed to migrate document to the latest version.'
+      );
+    }
+  }
+
+  /* ================= MULTINAMESPACE TYPES: CHECK NAMESPACES ================== */
+  let preflightCheckNamespacesResult: PreflightCheckNamespacesResult | undefined;
+
+  // handling multinamespace types
+
+  if (registry.isMultiNamespace(type)) {
+    logger.info('Verifying multi-namespace type');
+    preflightCheckNamespacesResult = await preflightHelper.preflightCheckNamespaces({
+      type,
+      id,
+      namespace,
+    });
+  }
+  const existingNamespaces = preflightCheckNamespacesResult?.savedObjectNamespaces ?? [];
+
+  const authorizationResult = await securityExtension?.authorizeUpdate({
+    namespace,
+    object: { type, id, existingNamespaces },
+  });
+
+  /**
+   * We're now in a situation where, for multi-namespace types, we've got multiple results
```

Comment:
Ideally, we need to combine these. For a non-bulk update, the type is either multi-namespace or not.